### PR TITLE
Protect native allocated typed array from being GC

### PR DESCRIPTION
**Pull request by PatriceJiang**
ref https://forum.cocos.com/t/ios/76826, https://forum.cocos.com/t/ios-glteximage2d-exc-bad-access/74364

防止对象被提前GC

### CODE REVIEW

**Comment by PatriceJiang** [comment]
@minggo  麻烦有空review下

File: cocos/scripting/js-bindings/manual/jsb_cocos2dx_manual.cpp [hunk 1]
```diff
@@ -436,13 +436,15 @@ static bool js_CanvasRenderingContext2D_setCanvasBufferUpdatedCallback(se::State
                     se::ValueArray args;
                     args.resize(1);
                     ok &= Data_to_seval(larg0, &args[0]);
+                    args[0].toObject()->root();
```
Comment (minggo) [inline cocos/scripting/js-bindings/manual/jsb_cocos2dx_manual.cpp]: se::Value 的构造函数可以传入参数防止 GC

Comment (PatriceJiang) [reply cocos/scripting/js-bindings/manual/jsb_cocos2dx_manual.cpp]: https://github.com/cocos-creator/cocos2d-x-lite/blob/3e7ccddda6dd08cbdaa84f35bad81a33ddec87ad/cocos/scripting/js-bindings/jswrapper/Value.hpp#L141

https://github.com/cocos-creator/cocos2d-x-lite/blob/3e7ccddda6dd08cbdaa84f35bad81a33ddec87ad/cocos/scripting/js-bindings/jswrapper/Value.hpp#L148

感觉这里显式调用比较合适

Comment (PatriceJiang) [reply cocos/scripting/js-bindings/manual/jsb_cocos2dx_manual.cpp]: 改成在Data_to_seval中处理

File: cocos/scripting/js-bindings/manual/jsb_conversions.cpp [hunk 2]
```diff
@@ -2509,7 +2509,7 @@ bool Data_to_seval(const cocos2d::Data& v, se::Value* ret)
     else
     {
         se::HandleObject obj(se::Object::createTypedArray(se::Object::TypedArrayType::UINT8, v.getBytes(), v.getSize()));
-        ret->setObject(obj);
+        ret->setObject(obj, true);
```
Comment (minggo) [inline cocos/scripting/js-bindings/manual/jsb_conversions.cpp]: HandleObject 应该已经处理了 root, unroot 了吧。

Comment (PatriceJiang) [reply cocos/scripting/js-bindings/manual/jsb_conversions.cpp]: `HandleObject`在析构时进行了`unroot`, 过早了。 需要把`typed array`扩展到`se::Value* ret`的生命周期。

Comment (minggo) [reply cocos/scripting/js-bindings/manual/jsb_conversions.cpp]: 有两个疑问：
1. 这个文件里的所有函数都有这个问题。
2. 如果这样的话，那么就没必要创建 HandleObject 对象的吧。

Comment (PatriceJiang) [reply cocos/scripting/js-bindings/manual/jsb_conversions.cpp]: 1. 是的
2. HandleObject设计上可以 保护初始化过程中不被 GC

**Comment by minggo** [comment]
GC 的问题先这么临时处理吧。等使用了 V8 有了 scope 后就不会有这个问题了。不过 JSC 也是很奇葩，新创建的内存照理应该是放在新生代，怎么这么快就被 GC 了。

**Comment by jareguo** [comment]
这个问题也是这个地方引起的吗？ https://forum.cocos.com/t/jsobjectref-root-unroot-gc/74369

**Comment by PatriceJiang** [comment]
是的 @jareguo

**Comment by jareguo** [comment]
修复原生平台上 GC 可能导致崩溃的问题

**Comment by caochao** [comment]
赞一个

**Comment by haroel** [comment]
我们使用的creator 2.3.4上也有类似崩溃，尝试加root/unroot也还是会报出错
android v8

[ libcocos2djs.so ] 
v8::internal::GlobalHandles::ClearWeakness(unsigned long*)
:?
pc 0x00000000005a3a64
[ libcocos2djs.so ] 
se::Object::root()
D:/xxxxxx/../../../xxxxxx/Cocos2dx_2/cocos/scripting/js-bindings/jswrapper/v8/Object.cpp:714
pc 0x00000000005a1098
[ libcocos2djs.so ] 
se::Value::setObject(se::Object*, bool)
D:/xxxxxx/../../../xxxxxx/Cocos2dx_2/cocos/scripting/js-bindings/jswrapper/Value.cpp:400
pc 0x00000000005a0a10
[ libcocos2djs.so ] 
se::Value::operator=(se::Value const&)
D:/xxxxxx/../../../xxxxxx/Cocos2dx_2/cocos/scripting/js-bindings/jswrapper/Value.cpp:198
pc 0x0000000000593fa8
[ libcocos2djs.so ] 
std::__ndk1::__compressed_pair_elem<jsb_global_load_image(std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&, se::Value const&)::$_4::operator()(std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&, unsigned char*, int, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&) const::{lambda(int)#1}::operator()(int) const::{lambda()#1}, 0, false>::__compressed_pair_elem<{lambda(int)#1} const&, 0ul>(std::__ndk1::piecewise_construct_t, std::__ndk1::tuple<{lambda(int)#1} const&>, std::__ndk1::__tuple_indices<0ul>)
D:/xxxxxx/../../../xxxxxx/Cocos2dx_2/cocos/scripting/js-bindings/manual/jsb_global.cpp:861
pc 0x00000000005936e0
[ libcocos2djs.so ] 
__compressed_pair<const (lambda at D:/xxxxxx/../../../xxxxxx/Cocos2dx_2/cocos/scripting/js-bindings/manual/jsb_global.cpp:861:86) &, std::__ndk1::allocator<(lambda at D:/xxxxxx/../../../xxxxxx/Cocos2dx_2/cocos/scripting/js-bindings/manual/jsb_global.cpp:861:86)> &&>
F:/androidSDK\ndk\android-ndk-r21d\build\..\sources\cxx-stl\llvm-libc++\include/memory:2256
pc 0x000000000053a068
[ libcocos2djs.so ] 
__value_func
F:/androidSDK\ndk\android-ndk-r21d\build\..\sources\cxx-stl\llvm-libc++\include/functional:1802
pc 0x0000000000539e48
[ libcocos2djs.so ] 
cocos2d::Scheduler::update(float)
D:/xxxxxx/../../../xxxxxx/Cocos2dx_2/cocos/base/CCScheduler.cpp:513
pc 0x000000000022b600
[ libcocos2djs.so ] 
Java_org_cocos2dx_lib_Cocos2dxRenderer_nativeRender
D:/xxxxxx/../../../xxxxxx/Cocos2dx_2/cocos/platform/android/jni/JniImp.cpp:252

**Pull request merged**